ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala

```diff
@@ -146,6 +142,157 @@ object KeyValueCommitting {
     }
   }
 
+  private def processPackageUpload(
+      entryId: DamlLogEntryId,
+      recordTime: Timestamp,
+      packageUploadEntry: DamlPackageUploadEntry,
+      inputState: Map[DamlStateKey, Option[DamlStateValue]]
+  ): (DamlLogEntry, Map[DamlStateKey, DamlStateValue]) = {
+    val submissionId = packageUploadEntry.getSubmissionId
+    def tracelog(msg: String): Unit =
+      logger.trace(
+        s"processPackageUpload[entryId=${prettyEntryId(entryId)}, submId=$submissionId]: $msg")
+
+    val archives = packageUploadEntry.getArchivesList.asScala
+    val packagesValidityResult =
+      archives.foldLeft[(Boolean, String)]((true, ""))(
+        (acc, archive) =>
+          if (archive.getPayload.isEmpty) (false, acc._2 ++ s"empty package '${archive.getHash}';")
+          else acc)
+
+    if (packagesValidityResult._1) {
```

Comment:
swap this so we can read it more naturally: check for errors, handle the error, otherwise continue.